cpp/src/c_api/legacy_fa2.cpp

```diff
@@ -0,0 +1,248 @@
+  bool outbound_attraction_distribution_{};
+  bool lin_log_mode_{};
+  bool prevent_overlapping_{};
+  const double edge_weight_influence_{};
+  const double jitter_tolerance_{};
+  bool barnes_hut_optimize_{};
+  const double barnes_hut_theta_{};
+  const double scaling_ratio_{};
+  bool strong_gravity_mode_{};
+  const double gravity_{};
+  bool verbose_{};
+  bool do_expensive_check_{};
+  cugraph::c_api::cugraph_type_erased_device_array_t* result_{};;
+
+  force_atlas2_functor(::cugraph_resource_handle_t const* handle,
+                       ::cugraph_graph_t* graph,
+                       const int max_iter,
+                       ::cugraph_type_erased_device_array_view_t* x_start,
+                       ::cugraph_type_erased_device_array_view_t* y_start,
+                       bool outbound_attraction_distribution,
+                       bool lin_log_mode,
+                       bool prevent_overlapping,
+                       const double edge_weight_influence,
+                       const double jitter_tolerance,
+                       bool barnes_hut_optimize,
+                       const double barnes_hut_theta,
+                       const double scaling_ratio,
+                       bool strong_gravity_mode,
+                       const double gravity,
+                       bool verbose,
+                       bool do_expensive_check)
+    : abstract_functor(),
+      handle_(*reinterpret_cast<cugraph::c_api::cugraph_resource_handle_t const*>(handle)->handle_),
+      graph_(reinterpret_cast<cugraph::c_api::cugraph_graph_t*>(graph)),
+      max_iter_(max_iter),
+      x_start_(
+        reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t*>(x_start)),
+      y_start_(
+        reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t*>(y_start)),
+      outbound_attraction_distribution_(outbound_attraction_distribution),
+      lin_log_mode_(lin_log_mode),
+      prevent_overlapping_(prevent_overlapping),
+      edge_weight_influence_(edge_weight_influence),
+      jitter_tolerance_(jitter_tolerance),
+      barnes_hut_optimize_(barnes_hut_optimize),
+      barnes_hut_theta_(barnes_hut_theta),
+      scaling_ratio_(scaling_ratio),
+      strong_gravity_mode_(strong_gravity_mode),
+      gravity_(gravity),
+      verbose_(verbose),
+      do_expensive_check_(do_expensive_check)
+  {
+  }
+
+  template <typename vertex_t,
+            typename edge_t,
+            typename weight_t,
+            typename edge_type_type_t,
+            bool store_transposed,
+            bool multi_gpu>
+  void operator()()
+  {
+    if constexpr (!cugraph::is_candidate<vertex_t, edge_t, weight_t>::value) {
+      unsupported();
+    } else if constexpr (multi_gpu) {
+      unsupported();
+    } else if constexpr (!std::is_same_v<edge_t, int32_t>) {
+      unsupported();
+    } else {
+      // FIXME: Does FA2 expects store_transposed == false
```

Comment:
FA2 is expecting a `legacy::GraphCOOView`, which you're creating below.  there's no reason to transpose the graph, you should be able to use it exactly as is.

I'd take out this logic.  You can then remove the annotation in the C API documentation that the graph object might be modified... because in the case of this algorithm it would not be.